### PR TITLE
*: txn add SetMemBufCap, use to reduce memory allocations.

### PR DESCRIPTION
**Pull request by winkyao**
run bench: https://github.com/pingcap/tidb-lightning/tree/cenqichao/kv-encoder-bench

use round 50  values batch 32 * 1024:

```
./bin/bench-encoder -r 50 -b 32768 -s 0
```

got results:

```
master:
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 87.632 sec

16k, temporary BufferStore membufCap set to len(t.indiecs):
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 74.639 sec

16k, temporary BufferStore membufCap set to len(t.indiecs) * 2:
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 72.369 sec

32k, len(t.indiecs) * 2:
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 68.688 sec

64k, len(t.indiecs) * 2:
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 72.058 sec

32k, len(t.indiecs) * 4:
[direct] kv num = 19660800 / kv size = 2179891200
[direct] cost avg = 74.507 sec
```
with ImportingTxnMembufCap set to 32k, and bs.membufCap in AddRecord set to len(t.indiecs) * 2, has about 21.6% improvement.

### CODE REVIEW

**Comment by winkyao** [comment]
@shenli @coocood PTAL

File: table/tables/tables.go [hunk 1]
```diff
@@ -212,7 +212,7 @@ func (t *Table) FirstKey() kv.Key {
 // Length of `oldData` and `newData` equals to length of `t.WritableCols()`.
 func (t *Table) UpdateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, touched []bool) error {
 	txn := ctx.Txn()
-	bs := kv.NewBufferStore(txn)
+	bs := kv.NewBufferStore(txn, kv.SmallTxnMembufCap)
```
Comment (shenli) [inline table/tables/tables.go]: Why not use (len(t.indices) + 1)*2?

File: table/tables/tables.go [hunk 2]
```diff
@@ -323,7 +323,7 @@ func (t *Table) AddRecord(ctx context.Context, r []types.Datum, skipHandleCheck
 	}
 
 	txn := ctx.Txn()
-	bs := kv.NewBufferStore(txn)
+	bs := kv.NewBufferStore(txn, len(t.indices)*2)
```
Comment (shenli) [inline table/tables/tables.go]: `len(t.indices) + 1` is enough.

File: session.go [hunk 3]
```diff
@@ -135,6 +135,14 @@ type session struct {
 	statsCollector *statistics.SessionStatsCollector
 }
 
+func (s *session) getMembufCap() int {
+	if s.sessionVars.ImportingData {
+		return kv.ImportingTxnMembufCap
```
Comment (shenli) [inline session.go]: Can we make it a session variable? Other users could benefit from it.

**Comment by coocood** [comment]
@winkyao 
The capacity is the size of byte slice, not the number of key/value pairs.

**Pull request closed without merging**